### PR TITLE
Allow multiple noinline instances with different arguments.

### LINKED ISSUE #6646: Allow multiple patterns for noinline
I have multiple function patterns that I would like to exclude from inlining. Unfortunately, the noinline pass only allows a single pattern, which does not allow any conditionals. It would be great if there were a way to specify multiple patterns, either by extending the pattern syntax or by allowing multiple instances of the noinline pass.

I can provide a PR for either of these options if desired 😏

### PR DESCRIPTION
**Pull request by DirtyHairy**
Allow multiple `no-inline` - passes as suggested in #6646 .

### CODE REVIEW

File: src/passes/NoInline.cpp [hunk 1]
```diff
@@ -48,8 +48,10 @@ struct NoInline : public Pass {
   NoInline(NoInlineMode mode) : mode(mode) {}
 
   void run(Module* module) override {
-    std::string pattern = getPassOptions().getArgument(
-      name, "Usage usage:  wasm-opt --" + name + "=WILDCARD");
+    std::string pattern =
+      passArg ? *passArg
+              : getPassOptions().getArgument(
```
Comment (kripken) [inline src/passes/NoInline.cpp:53]: I wonder if we can simplify this? Specifically that it accepts two mechanisms seems a little more complex than it would ideally be.

In fact `--passname=ARG` sets both of them, so I think we only need to check the former? But I might be missing something.

File: src/pass.h [hunk 2]
```diff
@@ -47,6 +51,7 @@ struct PassRegistry {
   std::vector<std::string> getRegisteredNames();
   std::string getPassDescription(std::string name);
   bool isPassHidden(std::string name);
+  bool doesPassAllowMultipleInstancesWithArgs(std::string name);
```
Comment (kripken) [inline src/pass.h:54]: Can we make this the behavior of all passes? I can't think of a problem that would occur, and that seems simpler, but again I might be missing something.

**Review by kripken — COMMENTED** [review]
Nice!

File: src/passes/NoInline.cpp (hunk 1 above)
Comment (DirtyHairy) [reply src/passes/NoInline.cpp:53]: The reason why I kept the global mechanism is `--pass-arg`.

File: src/pass.h (hunk 2 above)
Comment (DirtyHairy) [reply src/pass.h:54]: Sure. I introduced the flag only for strict backwards compatibility for the other passes (bail out if the pass is specified multiple times with different arguments), but I'd be more than happy to remove it 😏

File: src/passes/NoInline.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/NoInline.cpp:53]: Ah, yes, I agree we want the global mechanism too. And that makes sense: there can be a global flag that multiple passes (even of different operations, not just instances of the same Pass) care about, but also each pass instance can have it's own. But here, I think this can always read `passArg` rather than check `passArg` and maybe read `getPassOptions().getArgument(..)`.

File: src/pass.h (hunk 2 above)
Comment (kripken) [reply src/pass.h:54]: Ok good, I don't think we need to bother with the case of passes passed different arguments before now (that was never meant to work). So if no tests break, then let's make that change.

Comment (DirtyHairy) [reply src/pass.h:54]: OK. Before I go ahead and make those changes, just to make sure we have the same picture of how the result should look like: All passes will use their own private value of the option that shares their name, and only options that have a different name and thus are set via `--pass-arg` will be read from the global map.

Comment (kripken) [reply src/pass.h:54]: Yes, that sounds right.

Comment (DirtyHairy) [reply src/pass.h:54]: Alright, I'll make the changes.

Comment (DirtyHairy) [reply src/pass.h]: Done. There were a few tests that were affected by the change to the semantics of `--pass-arg`  (cannot be used to set the "primary" argument anymore). If this is too invasive we can add the fallback to the global argument back again 😏

Comment (kripken) [reply src/pass.h]: Is this comment no longer valid?

https://github.com/WebAssembly/binaryen/blob/a12f18b1d6bfdcb565f755822e44594666a6cd0e/src/tools/optimization-options.h#L310-L317

It is not updated in this PR. If that behavior changes, please update it. However, I think we can keep that old behavior? I don't see a reason for a breaking change here (aside from corner cases of running a pass more than once with different arguments, which this PR should be improving).

Comment (DirtyHairy) [reply src/pass.h]: That's what I referred to with

> [...] and only options that have a different name and thus are set via --pass-arg will be read from the global map.

above 😏  Avoiding this breaking change was the reason why I added the fallback to the global option for in my first version of the PR. If that's fine with you I'll just add the fallback to the global version of the argument back to `getArgument` & friends, this will restore the old behaviour.

Comment (kripken) [reply src/pass.h]: Ah, I think I wasn't clear enough before, or I've misunderstood you. To make sure we are on the same page, the issue is that we need the fallback from the pass's argument to the global argument, because of the two forms `--foopass=FOO` and `--pass-arg=foopass@FOO`?

I think we can get around that. It looks like optimization-options handles the first form, and tool-options the second, but if they were coordinated this could be fixed, and passes would only read from the one place.

Concretely, I'm suggesting this:

* When we see `--foopass=FOO` then we check add it to the pass instance as an argument.
* When we see `--pass-arg=foopass@FOO` then we check if there is a pass for it.
  * If there is, we write it to that pass instance as an argument. If there is more than one appearance of the pass, set it for them all.
  * If there is no such pass, write it to the global store.

There is no need for a fallback here, since we always write to the pass instance when it has the name of a pass, and always to the global store when not.

That means some more processing in tool-options/optimization-options, but I think it's worth it. It would be simpler, and also backwards compatible except for corner cases with multiple invocations of the same pass (which I am ok with altering).

Comment (DirtyHairy) [reply src/pass.h]: Yeah, I think we managed to misunderstand each other 😏 

> To make sure we are on the same page, the issue is that we need the fallback from the pass's argument to the global argument, because of the two forms --foopass=FOO and --pass-arg=foopass@FOO?

Exactly.

I think your suggestion would work well, although I'd suggest a minor change:

* When we see `--foopass=FOO` we set the argument on the pass instance
* When we see `--pass-arg=foopass@FOO` we check the pass list for `foopass`
   - If we find an instance of `foopass` we set the argument there and do not continue to look for other, earlier instances
   - If there is no instance we store the argument in the global list, take it from there when we add `foopass` after     that and remove it from the global list afterwards

This way, `--pass-arg=foopass` will always affect the previous instance of `foopass` on the command line, except when it occurs before the first `--foopass`, in which case it will affect the first instance. To me, this is a little more consistent with how I would expect the UI to behave (instead of a single `--pass-arg=foopass` overriding all previous instances).

Comment (kripken) [reply src/pass.h]: Interesting. I like that it affects just one instance. Maybe we can make that even simpler, though: `--pass-arg=foopass@FOO` can always "attach" to the last instance, as you suggested, but also (1) it is an error if that pass already has an arg, and (2) it is an error if there is no pass instance before. Those avoid some possible ambiguity I think?

Comment (DirtyHairy) [reply src/pass.h]: I like that even better 😏 I'll make the necessary changes.

Comment (DirtyHairy) [reply src/pass.h]: OK, done. I also rebased my code on the current main head. If we agree that this is the behaviour we want to stick with I can also add a few tests.

File: src/binaryen-c.cpp [hunk 3]
```diff
@@ -5451,7 +5451,9 @@ void BinaryenModuleRunPasses(BinaryenModuleRef module,
   PassRunner passRunner((Module*)module);
   passRunner.options = globalPassOptions;
   for (BinaryenIndex i = 0; i < numPasses; i++) {
-    passRunner.add(passes[i]);
+    passRunner.add(passes[i]), globalPassOptions.arguments.count(passes[i]) > 0
+                                 ? globalPassOptions.arguments[passes[i]]
+                                 : std::optional<std::string>();
```
Comment (kripken) [inline src/binaryen-c.cpp:5456]: I don't understand this line... it looks like
```cpp
A, B ? C : D;
```
where `A` has an effect (it adds the pass) but `B` just counts something and `C` and `D` have no side effects, so this looks equivalent to
```cpp
A;
```
which is
```cpp
passRunner.add(passes[i])
```
Perhaps this was meant to be
```cpp
passRunner.add(passes[i], ....);
```
(rather than to use the comma operator after closing the `add`)?

Comment (DirtyHairy) [reply src/binaryen-c.cpp]: Of course, you're right, I messed up the braces. Fixed.

File: src/binaryen-c.cpp [hunk 4]
```diff
@@ -5451,7 +5451,10 @@ void BinaryenModuleRunPasses(BinaryenModuleRef module,
   PassRunner passRunner((Module*)module);
   passRunner.options = globalPassOptions;
   for (BinaryenIndex i = 0; i < numPasses; i++) {
-    passRunner.add(passes[i]);
+    passRunner.add(passes[i],
+                   globalPassOptions.arguments.count(passes[i]) > 0
+                     ? globalPassOptions.arguments[passes[i]]
+                     : std::optional<std::string>());
```
Comment (kripken) [inline src/binaryen-c.cpp:5457]: Does it compile with `{}` or `std::nullopt` here? I'd prefer them in that order if possible, but if the compiler says no, we can leave it...

File: src/pass.h [hunk 5]
```diff
@@ -39,12 +39,14 @@ struct PassRegistry {
   using Creator = std::function<Pass*()>;
 
   void registerPass(const char* name, const char* description, Creator create);
+  ;
```
Comment (kripken) [inline src/pass.h:42]:
```suggestion

```

File: src/pass.h [hunk 6]
```diff
@@ -322,9 +327,8 @@ struct PassRunner {
   }
 
   // Add a pass using its name.
-  void add(std::string passName) {
-    doAdd(PassRegistry::get()->createPass(passName));
-  }
+  void add(std::string passName,
+           std::optional<std::string> passArg = std::optional<std::string>());
```
Comment (kripken) [inline src/pass.h:331]:
```suggestion
           std::optional<std::string> passArg = std::nullopt);
```
This should compile as we use this pattern elsewhere.

File: src/passes/pass.cpp [hunk 7]
```diff
@@ -45,7 +45,7 @@ void PassRegistry::registerPass(const char* name,
                                 const char* description,
                                 Creator create) {
   assert(passInfos.find(name) == passInfos.end());
-  passInfos[name] = PassInfo(description, create);
+  passInfos[name] = PassInfo(description, create, false);
```
Comment (kripken) [inline src/passes/pass.cpp:48]: Why is this needed? Isn't false the default?

File: src/passes/pass.cpp [hunk 8]
```diff
@@ -73,6 +73,16 @@ std::vector<std::string> PassRegistry::getRegisteredNames() {
...
 }
 
+bool PassRegistry::containsPass(const std::string& name) {
+  for (auto& [passName, _] : passInfos) {
+    if (passName == name) {
+      return true;
+    }
+  }
+
+  return false;
```
Comment (kripken) [inline src/passes/pass.cpp:83]:
```suggestion
  return passInfos.count(name) > 0;
```

File: src/passes/pass.cpp [hunk 9]
```diff
@@ -1025,4 +1044,28 @@ bool PassRunner::shouldPreserveDWARF() {
   return true;
 }
 
+bool Pass::hasArgument(const std::string& key) {
+  return (key == name) ? passArg.has_value()
```
Comment (kripken) [inline src/passes/pass.cpp:1048]:
```suggestion
  // An argument with the name of the pass is stored on the instance. Other arguments are in the global storage.
  return key == name ? passArg.has_value()
```

File: src/tools/optimization-options.h [hunk 10]
```diff
@@ -315,18 +317,38 @@ struct OptimizationOptions : public ToolOptions {
...
-          passes.push_back(p);
+
+          passes.push_back(info);
         },
         PassRegistry::get()->isPassHidden(p));
     }
   }
 
+  void addPassArg(const std::string& key, const std::string& value) override {
+    for (auto iPass = passes.rbegin(); iPass != passes.rend(); iPass++) {
```
Comment (kripken) [inline src/tools/optimization-options.h:332]: `iPass => i` would match our conventions better.

File: src/tools/optimization-options.h [hunk 11]
```diff
@@ -315,18 +317,38 @@ struct OptimizationOptions : public ToolOptions {
...
           }
-          passes.push_back(p);
+
+          passes.push_back(info);
         },
         PassRegistry::get()->isPassHidden(p));
     }
   }
 
+  void addPassArg(const std::string& key, const std::string& value) override {
```
Comment (kripken) [inline src/tools/optimization-options.h:331]: This could benefit from a comment as to what it is doing overall (looking for the last appearance etc.).

**Review by kripken — COMMENTED** [review]
Nice work!

**Comment by kripken** [comment]
Also, the behavior we decided on should be documented somewhere - did I miss that in the code?

File: src/passes/pass.cpp (hunk 7 above)
Comment (DirtyHairy) [reply src/passes/pass.cpp:48]: Indeed. This is a leftover from when I had added a second default argument. Removed.

File: src/passes/pass.cpp (hunk 8 above)
Comment (DirtyHairy) [reply src/passes/pass.cpp:83]: That would've been too easy 😏

File: src/tools/optimization-options.h (hunk 11 above)
Comment (DirtyHairy) [reply src/tools/optimization-options.h:335]: I added a few comments to the function.

**Comment by DirtyHairy** [comment]
> Also, the behavior we decided on should be documented somewhere - did I miss that in the code?

No, there isn't anything yet, but I agree, some documentation would be nice and helpful. Do you have a pointer what a good place would be?

**Comment by kripken** [comment]
The description of `--pass-arg` would be one user-facing place. For internal code documentation, in `pass.h` where the pass has its argument could be a good place.

**Comment by DirtyHairy** [comment]
Thanks 😏 I've added a few comments.

File: src/tools/tool-options.h [hunk 12]
```diff
@@ -126,7 +126,9 @@ struct ToolOptions : public Options {
       .add("--pass-arg",
            "-pa",
            "An argument passed along to optimization passes being run. Must be "
-           "in the form KEY@VALUE",
+           "in the form KEY@VALUE. If a pass --somepass is specified "
+           "multiple times, --pass-arg=somepass will apply to the closest "
+           "instance to the left. All other options apply to all instances.",
```
Comment (kripken) [inline src/tools/tool-options.h:131]: This should also say "if KEY is the name of a pass, then [..]". Overall maybe:

> in the form KEY@VALUE. If KEY is the name of a pass then it applies to the closest instance of that pass before us. If KEY is not the name of a pass then it is a global option that applies to all pass instances that read it (later appearances of such global options override earlier ones).

**Review by kripken — COMMENTED** [review]
Great lgtm with one suggestion.

**Comment by kripken** [comment]
Oh, also please add something to `CHANGELOG.md`

**Comment by DirtyHairy** [comment]
> Oh, also please add something to `CHANGELOG.md`

Done!

File: src/tools/tool-options.h (hunk 12 above)
Comment (DirtyHairy) [reply src/tools/tool-options.h]: I like you suggestion better than mine and adopted it, although I removed the last part in parenthesis as the text already is quite long, especially when formatted into a column.

File: CHANGELOG.md [hunk 13]
```diff
@@ -38,6 +38,12 @@ Current Trunk
  - The build-time option to use legacy WasmGC opcodes is removed.
  - The strings in `string.const` instructions must now be valid WTF-8.
  - The `TraverseCalls` flag for `ExpressionRunner` is removed.
+ - Passes can now be specified several times on the command line. The "main"
```
Comment (kripken) [inline CHANGELOG.md:41]: Passes could already be specified more than once - what changes here is that each such instance can get its own argument?

How about

> Passes can now receive individual pass arguments, that is `--foo=A --foo=B` for a pass `foo` will run the pass twice (which was possible before) and will now run it first with argument `A` and second with `B`. `--pass-arg=foo@BAR` will now apply to the most recent `--foo` pass on the commandline, if `foo` is a pass (while global pass arguments - that are not the name of a pass - remain, as before, global for all passes).

File: src/tools/tool-options.h (hunk 12 above)
Comment (kripken) [reply src/tools/tool-options.h]: Looks good!

**Comment by kripken** [comment]
I started the tests here. I think for them to pass you need to update the help text outputs, which you can do by running `./scripts/update_help_checks.py` and committing that output.

File: CHANGELOG.md (hunk 13 above)
Comment (DirtyHairy) [reply CHANGELOG.md]: You're right, by bad. I like your version and have edited the changelog accordingly.

**Comment by DirtyHairy** [comment]
> I started the tests here. I think for them to pass you need to update the help text outputs, which you can do by running `./scripts/update_help_checks.py` and committing that output.

Done! I've also rebased on the current `main`.

**Review by kripken — APPROVED** [review]
Great!

Sorry this took a bunch of iterations, but it is a change to the commandline interface so I think it's important to get it as good as we can.

**Comment by DirtyHairy** [comment]
> Sorry this took a bunch of iterations, but it is a change to the commandline interface so I think it's important to get it as good as we can.

No worries, that's totally fine, thanks for merging!
